articles/qos_configurability.md

```diff
@@ -0,0 +1,379 @@
+        return qos.is_reliable();  // only accept reliable communication
+    },
+    "id1"
+  });
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  QosOverridingOptions { 
+    [] (const QoSProfile & qos) -> bool {
+      return !qos.is_reliable();  // only accept best effort communication
+    },
+    "id2"
+  });
+```
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:
+      publisher_id1:  # {entity_type}_{id}
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+            history_depth: 100
+             history: keep_last
+      publisher_id2:  # {entity_type}_{id}
+        'my/fully/qualified/topic/name/here': 
+          reliability: best_effort
+            history_depth: 100
+            history: keep_last
+```
+
+### Lowering the amount of declared parameters
+
+#### Hidden parameters
+
+Currently, ROS 2 has the concept of hidden topics and services.
+Those aren't listed in the cli tool commands, except when the user explicitly asks them to be included.
+
+Similarly, hidden parameters could be added.
+That will avoid showing a big list of all policies.
+
+#### Only declaring some parameters
+
+Instead of hidden parameters that correspond to the QoS policies, we could only declare parameters for the policies that the user want to be reconfigurable.
+That will lower the amount of declared parameters, but make publisher/subscription creation a bit more verbose.
+
+Example API:
+
+```cpp
+enum class QosPolicyKind {
+  Reliability,
+  ...,
+};
+
+/// Options that are passed in subscription/publisher constructor
+class QosOverridingOptions
+{
+    /// Constructor, allowing to declare all the "default" parameters. If `false` is passed, reconfiguring QoS is not allowed.
+    explicit QosOverridingOptions(bool declare_default_parameters, std::string id = "");
+    /// Implicit constructor, a parameter for each policy provided in the initializer list will be declared.
+    QosOverridingOptions(std::initializer_list<QosPolicyKind>, std::string id = "");
+    /// Same with a callback for accepting/rejecting the qos profile.
+    QosOverridingOptions(std::initializer_list<QosPolicyKind>, QosCallback callback, std::string id = "");
+    /// Declares the "default" parameters and pass a callback.
+    QosOverridingOptions(QosCallback callback, std::string id = "");
+};
+```
+
+```cpp
+// New method in Node class
+template<
+typename MessageT,
+typename AllocatorT = std::allocator<void>,
+typename PublisherT = rclcpp::Publisher<MessageT, AllocatorT>>
+std::shared_ptr<PublisherT>
+create_publisher(
+  const std::string & topic_name,
+  const rclcpp::QoS & qos,
+  const rclcpp::QosOverridingOptions & qos_options = QosOverridingOptions{false},
+  const PublisherOptionsWithAllocator<AllocatorT> & options = PublisherOptionsWithAllocator<AllocatorT>()
+);
+```
+
+```cpp
+// example usage
+node->create_publisher(
+    "my_topic",
+    user_qos_provided_in_code_1,
```

Comment:
I think this example, and the other examples below, have too many arguments.
```suggestion
```

Reply:
good catch!

Reply:
Fixed in https://github.com/ros2/design/pull/300/commits/2d5d786d4893727ff1f3214d34c43daff8cbb998.